submissions/src/main/kotlin/gov/cdc/prime/reportstream/submissions/controllers/SubmissionController.kt

```diff
@@ -0,0 +1,185 @@
+                "report_id" to reportId.toString(),
+                "status" to status
+            )
+            tableClient.createEntity(tableEntity.setProperties(tableProperties))
+            logger.info("Inserted report into table storage: reportId=$reportId")
+
+            // Create and publish custom event
+            val reportReceivedEvent = ReportReceivedEvent(
+                timeStamp = reportReceivedTime,
+                reportId = reportId,
+                parentReportId = reportId,
+                rootReportId = reportId,
+                headers = filterHeaders(headers),
+                sender = clientId!!,
+                senderIP = headers["x-azure-clientip"].toString(),
+                fileSize = headers["content-length"].toString(),
+                blobUrl = blobClient.blobUrl
+            )
+            logger.debug("Created ReportReceivedEvent")
+
+            // Log to Application Insights
+            telemetryClient.trackEvent(
+                "ReportReceivedEvent",
+                mapOf("event" to objectMapper.writeValueAsString(reportReceivedEvent)),
+                null)
+            telemetryClient.flush()
+            logger.info("Tracked ReportReceivedEvent with Application Insights")
+
+            val response =
+                CreationResponse(
+                    reportId,
+                    status,
+                    OffsetDateTime.now(),
+                )
+            logger.info("Report submission successful: reportId=$reportId")
+
+            return ResponseEntity(response, HttpStatus.CREATED)
+        } catch (e: Exception) {
+            logger.error("Error during report submission: ${e.message}", e)
+            return ResponseEntity("Internal Server Error: ${e.message}", HttpStatus.INTERNAL_SERVER_ERROR)
+        }
+    }
+
+    private fun filterHeaders(headers: Map<String, String>): Map<String, String> {
+        val headersToInclude = listOf("client_id", "Content-Type", "payloadName")
+        return headers.filter { it.key in headersToInclude }
+    }
+
+    private fun formBlobName(
+        reportId: UUID,
+        headers: Map<String, String>,
+    ): String {
+        val senderName = headers["client_id"]?.lowercase()
+        return when (headers["Content-Type"]?.lowercase()) {
+            "application/hl7-v2" -> "receive/$senderName/$reportId.hl7"
+            "application/fhir+ndjson" -> "receive/$senderName/$reportId.fhir"
+            else -> "receive/$senderName/$reportId"
+                // throw IllegalArgumentException("Unsupported content-type: $contentType")
+        }
+    }
+
+    private fun validateHeaders(clientId: String?): ResponseEntity<String>? {
+
+        if (clientId.isNullOrEmpty()) {
+            logger.warn("Missing required header: client_id")
+            return ResponseEntity.badRequest().body("Missing required header: client_id.")
```

Comment:
Nit: could also handle this via an exception handler rather than returning null.  Might make the flow a little clearer

Reply:
another good call. Added to exception handling class